phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala

```diff
@@ -0,0 +1,315 @@
+      |
+      |      onMessage = onMsg;
+      |      setTimeout(function() {
+      |        for (var i = 0; i < inMessages.length; ++i)
+      |          onMessage(inMessages[i]);
+      |        inMessages = null;
+      |      }, 0);
+      |    },
+      |    send: function(msg) {
+      |      if (websocket === null)
+      |        return; // we are closed already. ignore message
+      |
+      |      if (outMsgBuf !== null)
+      |        outMsgBuf.push(msg);
+      |      else
+      |        sendImpl(msg);
+      |    }
+      |  }
+      |}).call(this);""".stripMargin
+
+    MemVirtualBinaryFile.fromStringUTF8("comSetup.js", code)
+  }
+
+  def send(msg: String): Unit = synchronized {
+    state match {
+      case WaitingForServerToStart(sendQueue) =>
+        state = WaitingForServerToStart(msg :: sendQueue)
+
+      case AwaitingConnection(sendQueue) =>
+        state = AwaitingConnection(msg :: sendQueue)
+
+      case Connected =>
+        sendNow(msg)
+
+      case Closing =>
+        // Ignore
+    }
+  }
+
+  private def sendNow(msg: String): Unit = {
+    val fragParts = msg.length / MaxCharPayloadSize
+
+    for (i <- 0 until fragParts) {
+      val payload = msg.substring(
+          i * MaxCharPayloadSize, (i + 1) * MaxCharPayloadSize)
+      mgr.sendMessage("1" + payload)
+    }
+
+    mgr.sendMessage("0" + msg.substring(fragParts * MaxCharPayloadSize))
+  }
+
+  private def receiveFrag(frag: String): Unit = synchronized {
+    /* The fragments are accumulated in an instance-wide buffer in case
+     * receiving a non-first fragment times out.
+     */
+    fragmentsBuf ++= frag.substring(1)
+
+    frag.charAt(0) match {
+      case '0' =>
+        // Last fragment of a message, send it
+        val result = fragmentsBuf.result()
+        fragmentsBuf.clear()
+        onMessage(result)
+
+      case '1' =>
+        // There are more fragments to come; do nothing
+
+      case _ =>
+        throw new AssertionError("Bad fragmentation flag in " + frag)
+    }
+  }
+
+  def close(): Unit = synchronized {
+    val oldState = state
+    state = Closing
+
+    oldState match {
+      case WaitingForServerToStart(_) =>
+        mgr.stop()
+        // The underlying run will never start, so succeed now.
+        promise.trySuccess(())
+
+      case AwaitingConnection(_) =>
+        /* Do nothing. We need to allow the already running process to get to
+         * the point where it connects, otherwise we won't be able to
+         * gracefully stop it. onConnectionOpened() will take care of tearing
+         * down the manager.
+         */
+
+      case Connected =>
+        mgr.stop()
+
+      case Closing =>
+        // Ignore
+    }
+  }
+}
+
+object ComRun {
+  private final val MaxByteMessageSize = 32768 // 32 KB
+  private final val MaxCharMessageSize = MaxByteMessageSize / 2 // 2B per char
+  private final val MaxCharPayloadSize = MaxCharMessageSize - 1 // frag flag
```

Comment:
It is fine with me if you retain the whole fragmentation stuff, but if we are OK with upgrading to PhantomJS 2 and Jetty9, this is probably not necessary anymore.

Reply:
We could do that, but it's definitely out of the scope of this PR.